### ISSUE TITLE
Images not working

### ISSUE THREAD

**Issue by Ryuske** [impresspages/ImpressPages#423]
I just installed ImpressPages (4.0.13) and when I went to upload an image it uploaded fine, but the only place with the right URL was on the preview/resize image window. It doesn't show up on the site, or in the show all window.

They're looking for a URL like:
/file/2014/05/07/file.jpg

I also tried removing everything completely, installed version 4.0.12 and I had the same issue? Is there a configuration for this somewhere, or?

**Comment by XEuRoMAN** [comment]
Wow! It's been two years and nobody has fixed this annoying bug.
Making a pull request with some mistypes corrected.

**Comment by maskas** [comment]
Thanks for the pull request. How did you reproduce this issue?

**Comment by XEuRoMAN** [comment]
TL;DR
You can reproduce bug by disabling `mod_rewrite`.

I thought for a second that I'm lunatic... But no, probably another Windows quirk (I do a lot of "sleep" for my PC).
So, I installed fresh new ImpressPages and guess what, everything works (including https://github.com/impresspages/ImpressPages/issues/788#issuecomment-229802024). I haven't change anything in my Apache configuration. I'm using WAMP server and green tick under `rewrite_module` always was here.
To prove myself that I'm not lunatic, I tried to debug "why it's working" (worse than "why it's not working").
Code still not reaching my fixed part in `Ip/Internal/Repository/ReflectionModel.php` but with `mod_rewrite` enabled, reflection is created when there is request to that image `https://github.com/impresspages/ImpressPages/blob/ea121628e03f2a3b0d5764ac1c74427a65422197/Ip/Internal/Repository/Job.php#L38`.
When `mod_rewrite` is disabled you just get 404.

Still, don't understand why I had `mod_rewrite` disabled but there is my own problems.

**Comment by maskas** [comment]
Thanks. Somehow it breaks the test. Will see if the test needs to be updated.

**Comment by maskas** [comment]
Finally, I found time to investigate this. The pull request doesn't connect all the dots in my head. Disabling mod_rewrite doesn't reproduce the bug either. Have I missed something in this long conversation on how to make error 404 to happen? Here is what I do:

- turn off mod_rewrite in Apache
- add 'rewritesDisabled' => true to the config
- confirm URLs now are with "index.php" stating, that ImpressPages is working without mod_rewrite
- add an image widget to the page, upload new image, and select it
- I see new image on the page

**Comment by XEuRoMAN** [comment]
Umm, I was not used to adding `rewritesDisabled` to the config. Maybe that fixes the case when `mod_rewrite` is disabled. But this piece of code has bug for sure. As I saw, `$onDemand` is always `true`, so `if` condition never will be met (because of exclamation mark) and `is_file` is always `false` because file path is wrong.

**Comment by maskas** [comment]
If mod_rewrite is disabled, you have to state it in config. Otherwise, the
system can't work properly.

It is possible, that in some circumstances the path is wrong. And maybe
there is a bug. But I need exact instructions to reproduce it.

Automated tests catch, that old photos are not being removed properly when
proposed changes applied.







*Mangirdas Skripka*
mangirdas@impresspages.org
+37067128095

www.impresspages.org

Become our fan on Facebook <http://www.facebook.com/ImpressPages> and
follow us on Twitter <http://www.twitter.com/ImpressPages>

**Comment by riartem** [comment]
Hello! Now I have same issue with nginx. 
I use recommended nginx settings, and all works fine except this case.
When I navigate to /file/2016/09/18/share.png server returns default nginx 404 page. I think php script even not calling.
Hope for you help.
Sorry for my English (I am not native speaker).

**Comment by jankus** [comment]
Do nice urls actually work for you? Or do you see index.php in your urls?

**Comment by riartem** [comment]
@jankus yes, pretty url's works fine

**Comment by jankus** [comment]
Then, I believe, you need to go deeper into the code and check why URL rewrites fail to handle such URL on your server. Or is that file actually exists in that specific location?

If file doesn't exist system takes care and creates it. From that time such URL should be handled as normal static file.

**Comment by riartem** [comment]
@jankus After some experiments I found that if I calling /file/2016/09/18/share**.pn** all works fine: request passed to Ip/Internal/Repository/Job.php.
But when I calling /file/2016/09/18/share.pn**g** it just check local file and return 404. Maybe you can help me with more correct nginx configuration?

**Comment by jankus** [comment]
Something is wrong with static file handling. I'm not nginx expert. Try to search for answers why static file path is not rewritten and forwarded to php.

Try to set default config from here (https://www.impresspages.org/help/nginx) and see if it works. If it does then check the differences from your current configuration.

**Comment by riartem** [comment]
Yes, I am using standart recommended configuration from that url.

**Comment by riartem** [comment]
Problem solved, 
this rule broke config: 
location ~* .(jpg|jpeg|png|gif|ico|css|js)$ {
expires 7d;
}
After removing all works.
